examples_apps/run_gate_synthesis.py

```diff
@@ -0,0 +1,469 @@
+   of magnitude :math:`r_i` and phase :math:`\phi_i`, and
+
+-  :math:`\Phi(\lambda_i)` is a single mode non-Gaussian operation, in
+   this case chosen to be the Kerr interaction
+   :math:`\mathcal{K}(\kappa_i)=e^{i\kappa_i\hat{n}^2}` of strength
+   :math:`\kappa_i`.
+
+Hyperparameters
+---------------
+
+First, we must define the **hyperparameters** of our layer structure:
+
+-  ``cutoff``: the simulation Fock space truncation we will use in the
+   optimization. The TensorFlow backend will perform numerical
+   operations in this truncated Fock space when performing the
+   optimization.
+
+-  ``depth``: The number of layers in our variational quantum
+   circuit. As a general rule, increasing the number of layers (and
+   thus, the number of parameters we are optimizing over) increases the
+   optimizer's chance of finding a reasonable local minimum in the
+   optimization landscape.
+
+-  ``reps``: the number of steps in the optimization routine performing
+   gradient descent
+
+Some other optional hyperparameters include:
+
+-  The standard deviation of initial parameters. Note that we make a
+   distinction between the standard deviation of *passive* parameters
+   (those that preserve photon number when changed, such as phase
+   parameters), and *active* parameters (those that introduce or remove
+   energy from the system when changed).
+"""
+import numpy as np
+
+import strawberryfields as sf
+from strawberryfields.ops import *
+from strawberryfields.utils import operation
+
+# Cutoff dimension
+cutoff = 10
+
+# gate cutoff
+gate_cutoff = 4
+
+# Number of layers
+depth = 20
+
+# Number of steps in optimization routine performing gradient descent
+reps = 300
+
+# Learning rate
+lr = 0.015
+
+# Standard deviation of initial parameters
+passive_sd = 0.1
+active_sd = 0.001
+
+
+######################################################################
+# Note that, unlike in state learning, we must also specify a *gate
+# cutoff* :math:`d`. This restricts the target unitary to its action on a
+# :math:`d`-dimensional subspace of the truncated Fock space, where
+# :math:`d\leq D`, where :math:`D` is the overall simulation Fock basis
+# cutoff. As a result, we restrict the gate synthesis optimization to only
+# :math:`d` input-output relations.
+#
+# The layer parameters :math:`\vec{\theta}`
+# -----------------------------------------
+#
+# We use TensorFlow to create the variables corresponding to the gate
+# parameters. Note that each variable has shape ``(depth,)``, with each
+# individual element representing the gate parameter in layer :math:`i`.
+#
+
+import tensorflow as tf
+
+# We attempt to import the legacy version of tf.einsum to make sure that no
+# issues arise when computing with complex tensors (this is required due to a
+# known issue in TF version 2.1)
+try:
+    from tensorflow.python.ops.special_math_ops import _einsum_v1
+
+    tf.einsum = _einsum_v1
+except ImportError:
+    pass
```

Comment:
Tried skipping the use of `tf.einsum` by using a solution with `tf.matmul`, however, got no convergence on TF 2.1.

The attempted solution is a part of a former commit https://github.com/XanaduAI/strawberryfields/pull/374/commits/3ea096bf403aeefc3373d65ecd5a960cee92bb6a .

Reply:
@antalszava, I think this can be removed now. I've merge #377 into this branch, so it should converge with 2.1 (the state learning tutorial converges correctly with 2.1 after #377 was merged in).